File: Triage/bap-0.7-moflow/libtracewrap/libtrace/src/ocaml/trace_container.ml
(** $Id: trace_container.ml 6415 2012-06-29 01:42:57Z edmcman $
    Trace container implementation.

    Bugs:
    We keep track of things as Int64's, but ML's IO only uses ints.
*)

exception TraceException of string

type frame = Frame_piqi.frame

(* XXX: Auto-pull this from C++ header file *)
let default_frames_per_toc_entry = 10000L
and default_auto_finish = false
and default_arch = Arch.Bfd_arch_i386
and default_machine = 0L

(* Internal definitions *)
let magic_number = 7456879624156307493L
and magic_number_offset = 0L
and trace_version_offset = 8L
and bfd_arch_offset = 16L
and bfd_machine_offset = 24L
and num_trace_frames_offset = 32L
and toc_offset_offset = 40L
and first_frame_offset = 48L

let out_trace_version = 1L
and lowest_supported_version = 1L
and highest_supported_version = 1L

let write_i64 oc i64 =
  let output = BatIO.output_string () in
  let () = BatIO.write_i64 output i64 in
  let binary = BatIO.close_out output in
  output_string oc binary

let read_i64 ic =
  let s = String.create 8 in
  (* Read 8 bytes into s *)
  let () = really_input ic s 0 8 in
  let input = BatIO.input_string s in
  let i = BatIO.read_i64 input in
  let () = BatIO.close_in input in
  i

(** [foldn f i n] is f (... (f (f i n) (n-1)) ...) 0 *)
let rec foldn64 ?(t=0L) f i n =
  let (-) = Int64.sub in
  match n-t with
  | 0L -> f i n
  | _ when n>t -> foldn64 ~t f (f i n) (n-1L)
  | n when n == -1L -> i (* otags has trouble with '-1L' *)
  | w -> raise (Invalid_argument "negative index number in foldn64")

(* End helpers *)

class writer ?(arch=default_arch) ?(machine=default_machine) ?(frames_per_toc_entry = default_frames_per_toc_entry) ?(auto_finish=default_auto_finish) filename =
  (* Open the trace file *)
  let oc = open_out_bin filename in
  (* Seek to the first frame *)
  let () = LargeFile.seek_out oc first_frame_offset in
object(self)

  val mutable toc = []
  val mutable num_frames = 0L
  val frames_per_toc_entry = frames_per_toc_entry
  val auto_finish = auto_finish
  val mutable is_finished = false

  initializer Gc.finalise (fun self -> if not self#has_finished then self#finish) self

  method add (frame:frame) =
    if num_frames <> 0L && Int64.rem num_frames frames_per_toc_entry = 0L then
      (* Put a toc entry *)
      toc <- (LargeFile.pos_out oc) :: toc;

    let () = num_frames <- Int64.succ num_frames in

    (* Convert to string so we know length *)
    let s = Frame_piqi_ext.gen_frame frame `pb in
    let len = Int64.of_int (String.length s) in
    if len <= 0L then
      raise (TraceException "Attempt to add zero-length frame to trace");

    (* Write the length in binary *)
    let () = write_i64 oc len in

    let old_offset = LargeFile.pos_out oc in
    (* Finally write the serialized string out. *)
    let () = output_string oc s in

    (* Double-check our size. *)
    assert ((Int64.add old_offset len)
        = (LargeFile.pos_out oc));

  method finish =
    if is_finished then raise (TraceException "finish called twice");

    let toc_offset = LargeFile.pos_out oc in
    (* Make sure the toc is the right size. *)
    let () = assert ((num_frames = 0L) || (Int64.div (Int64.pred num_frames) frames_per_toc_entry) = Int64.of_int (List.length toc)) in
    (* Write frames per toc entry. *)
    let () = write_i64 oc frames_per_toc_entry in
    (* Write toc to file. *)
    let () = List.iter (fun offset -> write_i64 oc offset) (List.rev toc) in
    (* Now we need to write the magic number, number of trace frames,
       and the offset of field m at the start of the trace. *)
    (* Magic number. *)
    let () = LargeFile.seek_out oc magic_number_offset in
    let () = write_i64 oc magic_number in
    (* Trace version. *)
    let () = LargeFile.seek_out oc trace_version_offset in
    let () = write_i64 oc out_trace_version in
    (* CPU architecture. *)
    let () = LargeFile.seek_out oc bfd_arch_offset in
    (* Goodbye type safety! *)
    let () = write_i64 oc (Int64.of_int (Obj.magic arch)) in
    (* Machine type. *)
    let () = LargeFile.seek_out oc bfd_machine_offset in
    let () = write_i64 oc machine in
    (* Number of trace frames. *)
    let () = LargeFile.seek_out oc num_trace_frames_offset in
    let () = write_i64 oc num_frames in
    (* Offset of toc. *)
    let () = LargeFile.seek_out oc toc_offset_offset in
    let () = write_i64 oc toc_offset in
    (* Finally close the final and mark us as finished. *)
    let () = close_out oc in
    is_finished <- true

    method has_finished = is_finished

end

class reader filename =
  let ic = open_in_bin filename in
  (* Verify magic number *)
  let () = if read_i64 ic <> magic_number then
      raise (TraceException "Magic number is incorrect") in
  (* Trace version *)
  let trace_version = read_i64 ic in
  let () = if trace_version < lowest_supported_version ||
      trace_version > highest_supported_version then
    raise (TraceException "Unsupported trace version") in
  (* Read arch type, break type safety *)
  let archnum = read_i64 ic in
  let () = if not (archnum < (Int64.of_int (Obj.magic Arch.Bfd_arch_last))) then
      raise (TraceException "Invalid architecture") in
  let arch : Arch.bfd_architecture = Obj.magic (Int64.to_int archnum) in
  let machine = read_i64 ic in
  (* Read number of trace frames. *)
  let num_frames = read_i64 ic in
  (* Find offset of toc. *)
  let toc_offset = read_i64 ic in
  (* Find the toc. *)
  let () = LargeFile.seek_in ic toc_offset in
  (* Read number of frames per toc entry. *)
  let frames_per_toc_entry = read_i64 ic in
  (* Read each toc entry. *)
  let toc_size = Int64.div (Int64.pred num_frames) frames_per_toc_entry in
  let toc =
    let toc_rev = foldn64
      (fun acc n ->
        if n = 0L then acc else
          (read_i64 ic) :: acc
      ) [] toc_size in
    Array.of_list (List.rev toc_rev)
  in
  (* We should be at the end of the file now. *)
  let () = assert ((LargeFile.pos_in ic) = (LargeFile.in_channel_length ic)) in
  object(self)
    val mutable current_frame = 0L

    method get_num_frames = num_frames

    method get_frames_per_toc_entry = frames_per_toc_entry

    method get_arch = arch

    method get_machine = machine

    method get_trace_version = trace_version

    method seek frame_number =
      (* First, make sure the frame is in range. *)
      let () = self#check_end_of_trace_num frame_number "seek to non-existant frame" in

      (* Find the closest toc entry, if any. *)
      let toc_number = Int64.div frame_number frames_per_toc_entry in

      current_frame <- (match toc_number with
      | 0L -> let () = LargeFile.seek_in ic first_frame_offset in
              0L
      | _ -> let () = LargeFile.seek_in ic (Array.get toc (Int64.to_int (Int64.pred toc_number))) in
             Int64.mul toc_number frames_per_toc_entry);

      while current_frame <> frame_number do
        (* Read frame length and skip that far ahead. *)
        let frame_len = read_i64 ic in
        let () = LargeFile.seek_in ic (Int64.add (LargeFile.pos_in ic) frame_len) in
        current_frame <- Int64.succ current_frame
      done

    method get_frame : frame =
      let () = self#check_end_of_trace "get_frame on non-existant frame" in
      let frame_len = read_i64 ic in
      if (frame_len <= 0L) then
        raise (TraceException (Printf.sprintf "Read zero-length frame at offset %#Lx" (LargeFile.pos_in ic)));
      let buf = String.create (Int64.to_int frame_len) in
      (* Read the frame info buf. *)
      let () = really_input ic buf 0 (Int64.to_int frame_len) in
      let f = Frame_piqi.parse_frame (Piqirun.init_from_string buf) in
      let () = current_frame <- Int64.succ current_frame in
      f

    method get_frames requested_frames =
      let () = self#check_end_of_trace "get_frame on non-existant frame" in
      (* The number of frames we copy is bounded by the number of
         frames left in the trace. *)
      let 
          num_frames = min requested_frames (Int64.sub num_frames current_frame)
      in
      List.rev (foldn64 ~t:1L (fun l n -> self#get_frame :: l) [] num_frames)

    method end_of_trace =
      self#end_of_trace_num current_frame

    method private end_of_trace_num frame_num =
      if Int64.succ frame_num > num_frames then
        true
      else
        false

    method private check_end_of_trace_num frame_num msg =
      if self#end_of_trace_num frame_num then
        raise (TraceException msg)

    method private check_end_of_trace msg =
      if self#end_of_trace then
        raise (TraceException msg)

    initializer self#seek 0L
  end

File: BAP-0.7-moflow/libtracewrap/libtrace/src/cpp/trace.container.cpp
/**
 * $Id: trace.container.cpp 6650 2012-08-18 00:48:38Z edmcman $
 *
 * Implementation of trace container.
 */

#include "trace.container.hpp"
#include <stdio.h>
#include <iostream>
#include <string>

#define WRITE(x) { if (fwrite(&(x), sizeof(x), 1, ofs) != 1) { throw (TraceException("Unable to write to trace")); } }
#define READ(x) { if (fread(&(x), sizeof(x), 1, ifs) != 1) { throw (TraceException("Unable to read from trace")); } }
#ifdef _WIN32
typedef uint64_t traceoff_t;
#define SEEKNAME _fseeki64
#define SEEK(f,x) { if (SEEKNAME(f, x, SEEK_SET) != 0) { throw (TraceException("Unable to seek in trace to offset " + x)); } }
#define TELL(f) _ftelli64(f)
#else
typedef off_t traceoff_t;
#define SEEKNAME fseeko
#define SEEK(f,x) { if (SEEKNAME(f, x, SEEK_SET) != 0) { throw (TraceException("Unable to seek in trace to offset " + x)); } }
#define TELL(f) ftello(f)
#endif

namespace SerializedTrace {

  TraceContainerWriter::TraceContainerWriter(std::string filename,
                                             bfd_architecture arch,
                                             uint64_t machine,
                                             uint64_t frames_per_toc_entry_in,
                                             bool auto_finish_in) throw (TraceException)
    : num_frames (0),
      frames_per_toc_entry (frames_per_toc_entry_in),
      arch (arch),
      mach (machine),
      auto_finish (auto_finish_in),
      is_finished (false)
  {
    ofs = fopen(filename.c_str(), "wb");
    if (!ofs) { throw (TraceException("Unable to open trace file for writing")); }
    SEEK(ofs, first_frame_offset);
  }

  TraceContainerWriter::~TraceContainerWriter(void) throw () {

    /** Call finish if it has not been called already ANd if
        auto_finish is set. */
    if (!is_finished && auto_finish) {
      try {
        finish();
      }
      catch (std::exception const &e) {
        std::cerr << "Exception " << e.what() << " occured during TraceContainerWriter's auto-finish" << std::endl;
      }
    }
  }

  void TraceContainerWriter::add(frame &f) throw (TraceException) {
    /* Is is time for a toc entry? */
    if (num_frames > 0 && (num_frames % frames_per_toc_entry) == 0) {
      /* Yes.  Add the file offset where we will insert this frame to
         toc. */
      toc.push_back(TELL(ofs));
    }

    num_frames++;

    /* Serialize to string so we can get the length. */
    std::string s;
    if (!(f.SerializeToString(&s))) {
      throw (TraceException("Unable to serialize frame to ostream"));
    }

    /* Write the length before the frame. */
    uint64_t len = s.length();
    if (len == 0) {
      throw (TraceException("Attempt to add zero-length frame to trace"));
    }
    WRITE(len);

    /* Write the frame. */
    traceoff_t old_offset = TELL(ofs);
    if (old_offset == -1) {
      throw (TraceException("Unable to determine the current offset in the trace"));
    }

    if (fwrite(s.c_str(), 1, len, ofs) != len) {
      throw (TraceException("Unable to write frame to trace file"));
    }

    /* Double-check our size. */
    assert ((uint64_t)old_offset + len == (uint64_t)TELL(ofs));
  }

  void TraceContainerWriter::finish(void) throw (TraceException) {
    if (is_finished) {
      throw (TraceException("finish called twice"));
    }

    /* Save the offset where we will write the toc. */
    uint64_t toc_offset = TELL(ofs);
    if (toc_offset == -1) {
      throw (TraceException("Unable to determine the current offset in the trace"));
    }

    /* Make sure the toc is the right size. */
    assert ((num_frames == 0) || ((num_frames - 1) / frames_per_toc_entry) == toc.size());

    /* Write frames per toc entry. */
    WRITE(frames_per_toc_entry);

    /* Write each offset to the file. */
    for (std::vector<uint64_t>::size_type i = 0; i < toc.size(); i++) {
      WRITE(toc[i]);
    }

    /* Now we need to write the magic number, number of trace frames
       and the offset of field m at the start of the trace. */

    /* Magic number. */
    SEEK(ofs, magic_number_offset);
    WRITE(magic_number);

    /* Trace version. */
    SEEK(ofs, trace_version_offset);
    WRITE(out_trace_version);

    /* CPU architecture. */
    SEEK(ofs, bfd_arch_offset);
    uint64_t archt = (uint64_t) arch;
    WRITE(archt);

    /* Machine type. */
    SEEK(ofs, bfd_machine_offset);
    WRITE(mach);

    /* Numer of trace frames */
    SEEK(ofs, num_trace_frames_offset);
    WRITE(num_frames);

    /* Offset of toc. */
    SEEK(ofs, toc_offset_offset);
    WRITE(toc_offset);

    /* Finally, close the file and mark us as finished. */
    if (fclose(ofs)) {
      throw (TraceException("Unable to close trace file"));
    }
    is_finished = true;
  }

  bool TraceContainerWriter::has_finished(void) throw () {
    return is_finished;
  }

  TraceContainerReader::TraceContainerReader(std::string filename) throw (TraceException)
  {
    ifs = fopen(filename.c_str(), "rb");
    if (!ifs) { throw (TraceException("Unable to open trace for reading")); }

    /* Verify the magic number. */
    uint64_t magic_number_read;
    READ(magic_number_read);
    if (magic_number_read != magic_number) {
      throw (TraceException("Magic number not found in trace"));
    }

    READ(trace_version);
    if (trace_version > highest_supported_version ||
        trace_version < lowest_supported_version) {
      throw (TraceException("Unsupported trace version"));
    }

    uint64_t archt;
    READ(archt);
    arch = (bfd_architecture) archt;

    READ(mach);

    /* Read number of trace frames. */
    READ(num_frames);

    /* Find offset of toc. */
    uint64_t toc_offset;
    READ(toc_offset);

    /* Find the toc. */
    SEEK(ifs, toc_offset);

    /* Read number of frames per toc entry. */
    READ(frames_per_toc_entry);

    /* Read each toc entry. */
    for (int i = 0; i < ((num_frames - 1) / frames_per_toc_entry); i++) {
      uint64_t offset;
      READ(offset);
      toc.push_back(offset);
    }

    /* We should be at the end of the file now. */
    traceoff_t us = TELL(ifs);
    traceoff_t end = SEEKNAME(ifs, 0, SEEK_END);
    if (us != TELL(ifs) || end != 0) {
      throw(TraceException("The table of contents is malformed."));
    }

    /* Seek to the first frame. */
    seek(0);
  }

  TraceContainerReader::~TraceContainerReader(void) throw () {
    /* Nothing yet. */
  }

  uint64_t TraceContainerReader::get_num_frames(void) throw () {
    return num_frames;
  }

  uint64_t TraceContainerReader::get_frames_per_toc_entry(void) throw () {
    return frames_per_toc_entry;
  }

  bfd_architecture TraceContainerReader::get_arch(void) throw () {
    return arch;
  }

  uint64_t TraceContainerReader::get_machine(void) throw () {
    return mach;
  }

  uint64_t TraceContainerReader::get_trace_version(void) throw () {
    return trace_version;
  }

  void TraceContainerReader::seek(uint64_t frame_number) throw (TraceException) {
    /* First, make sure the frame is in range. */
    check_end_of_trace_num(frame_number, "seek() to non-existant frame");

    /* Find the closest toc entry, if any. */
    uint64_t toc_number = frame_number / frames_per_toc_entry;

    if (toc_number == 0) {
      current_frame = 0;
      SEEK(ifs, first_frame_offset);
    } else {
      current_frame = toc_number * frames_per_toc_entry;
      /* Use toc_number - 1 because there is no toc for frames [0,m). */
      SEEK(ifs, toc[toc_number - 1]);
    }

    while (current_frame != frame_number) {
      /* Read frame length and skip that far ahead. */
      uint64_t frame_len;
      READ(frame_len);
      SEEK(ifs, (uint64_t)TELL(ifs) + frame_len);
      current_frame++;
    }
  }

  std::auto_ptr<frame> TraceContainerReader::get_frame(void) throw (TraceException) {
    /* Make sure we are in bounds. */
    check_end_of_trace("get_frame() on non-existant frame");

    uint64_t frame_len;
    READ(frame_len);
    if (frame_len == 0) {
      throw (TraceException("Read zero-length frame at offset " + TELL(ifs)));
    }

    /* We really just want a variable sized array, but MS VC++ doesn't support C99 yet.
     *
     * http://stackoverflow.com/questions/5246900/enabling-vlasvariable-length-arrays-in-ms-visual-c
     */
    auto_vec<char> buf ( new char[frame_len] );

    /* Read the frame into buf. */
    if (fread(buf.get(), 1, frame_len, ifs) != frame_len) {
      throw (TraceException("Unable to read frame from trace"));
    }

    std::string sbuf(buf.get(), frame_len);

    std::auto_ptr<frame> f(new frame);
    if (!(f->ParseFromString(sbuf))) {
      throw (TraceException("Unable to parse from string"));
    }
    current_frame++;

    return f;
  }

  std::auto_ptr<std::vector<frame> > TraceContainerReader::get_frames(uint64_t requested_frames) throw (TraceException) {
    check_end_of_trace("get_frames() on non-existant frame");

    std::auto_ptr<std::vector<frame> > frames(new std::vector<frame>);
    for (uint64_t i = 0; i < requested_frames && current_frame < num_frames; i++) {
      frames->push_back(*(get_frame()));
    }

    return frames;
  }

  bool TraceContainerReader::end_of_trace(void) throw () {
    return end_of_trace_num(current_frame);
  }

  bool TraceContainerReader::end_of_trace_num(uint64_t frame_num) throw () {
    if (frame_num + 1 > num_frames) {
      return true;
    } else {
      return false;
    }
  }

  void TraceContainerReader::check_end_of_trace_num(uint64_t frame_num, std::string msg) throw (TraceException) {
    if (end_of_trace_num(frame_num)) {
      throw (TraceException(msg));
    }
  }

    void TraceContainerReader::check_end_of_trace(std::string msg) throw (TraceException) {
      return check_end_of_trace_num(current_frame, msg);
    }
};
